pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -10,27 +12,112 @@ import (
 	"k8s.io/kubernetes/pkg/serviceaccount"
 
 	scopeauthorizer "github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	osclient "github.com/openshift/origin/pkg/client"
 	oauthapi "github.com/openshift/origin/pkg/oauth/api"
 	"github.com/openshift/origin/pkg/oauth/registry/oauthclient"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 const (
 	OAuthRedirectURISecretAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirecturi."
 	OAuthWantChallengesAnnotationPrefix    = "serviceaccounts.openshift.io/oauth-want-challenges"
+
+	OAuthRedirectModelAnnotationPrefix              = "serviceaccounts.openshift.io/oauth-redirectmodel."
+	OAuthRedirectModelAnnotationURIPrefix           = OAuthRedirectModelAnnotationPrefix + "uri."
+	OAuthRedirectModelAnnotationResourcePrefix      = OAuthRedirectModelAnnotationPrefix + "resource."
+	OAuthRedirectModelAnnotationURISchemePrefix     = OAuthRedirectModelAnnotationURIPrefix + "scheme."
+	OAuthRedirectModelAnnotationURIPortPrefix       = OAuthRedirectModelAnnotationURIPrefix + "port."
+	OAuthRedirectModelAnnotationURIPathPrefix       = OAuthRedirectModelAnnotationURIPrefix + "path."
+	OAuthRedirectModelAnnotationURIHostPrefix       = OAuthRedirectModelAnnotationURIPrefix + "host."
+	OAuthRedirectModelAnnotationResourceKindPrefix  = OAuthRedirectModelAnnotationResourcePrefix + "kind."
+	OAuthRedirectModelAnnotationResourceNamePrefix  = OAuthRedirectModelAnnotationResourcePrefix + "name."
+	OAuthRedirectModelAnnotationResourceGroupPrefix = OAuthRedirectModelAnnotationResourcePrefix + "group."
+
+	RouteKind = "Route"
 )
 
+var modelPrefixes = []string{
+	OAuthRedirectModelAnnotationURISchemePrefix,
+	OAuthRedirectModelAnnotationURIPortPrefix,
+	OAuthRedirectModelAnnotationURIPathPrefix,
+	OAuthRedirectModelAnnotationURIHostPrefix,
+	OAuthRedirectModelAnnotationResourceKindPrefix,
+	OAuthRedirectModelAnnotationResourceNamePrefix,
+	OAuthRedirectModelAnnotationResourceGroupPrefix,
+}
+
+type modelsToURIsFunc func(modelList []model) []redirectURI
```

Comment:
could this be something that would take a set of names for a particular groupkind, and return the components for the objects with those names? something like:
```
func uriComponentsFromRoutes(names sets.String) map[string][]uriComponents {
   var routes []api.Route 
   if len(names) == 1 {
     ... look up one route ...
   }else {
     ... list routes ...
   }

   retval := map[string][]uriComponents{}
   for _, route := range routes {
     if names.Has(route.Name) {
       retval[route.Name] = uriComponentsFromRoute(route)
     }
   }
   return retval
}

func uriComponentsFromRoute(route api.Route) []uriComponents {
  scheme := "http"
  if route.TLS != nil {
    scheme = "https"
  }

  retval := []uriComponents{}
  for _, ingress := route.Status.Ingress {
    if len(ingress.Host) > 0 && ... {
      retval = append(retval, uriComponents{Scheme: scheme, Host: ingress.Host}
    }
  }
  return retval
}

```